src/conduit/components/forms.cljs

```diff
@@ -0,0 +1,97 @@
+(ns conduit.components.forms
+  (:require [rum.core :as rum]
+            [clojure.string :refer [join]]
+            [citrus.core :as citrus]))
+
+(def initial-form-state {:data {} :errors {}})
+
+(def form-state (atom initial-form-state))
+
+(def email-regex #"[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?")
+
+(defn not-empty? [v]
+  (cond
+    (nil? v) false
+    (= "" v) false
+    :else true))
+
+(defn email? [v]
+  (not (nil? (re-matches email-regex (str v)))))
+
+(defn validate! [form-data form-errors validators]
+  (let [valid? (atom true)]
+    (doseq [[k v] validators]
+      (reset! form-errors (assoc @form-errors k nil))
+      (doall (->> v
+                  (map (fn [e]
+                         (let [[validator err-msg] e
+                               value (k @form-data)]
+                           (when (not (validator value))
+                             (reset! valid? false)
+                             (reset! form-errors (assoc @form-errors k (vec (conj (k @form-errors) err-msg)))))))))))
+    @valid?))
+
+(rum/defc InputErrors [errors input-key]
+  (when-let [input-errors (get-in errors [input-key])]
+    [:ul.error-messages
+     (map (fn [e] [:li e]) input-errors)]))
+
+(rum/defc ServerErrors < rum/reactive [r]
+  (when-let [server-errors (rum/react (citrus/subscription r [:user :errors]))]
+    [:ul.error-messages
+     (map (fn [[k v]]
+            [:li {:key k}
+             (str (name k) " "
+                  (if (vector? v) (join ", " v) v))])
+          server-errors)]))
+
+(rum/defc Input < rum/reactive
+  [{:keys [placeholder input-key input-type validators form-data form-errors]}]
+  (let [input-validators (select-keys validators [input-key])
+        validate-input! #(validate! form-data form-errors input-validators)]
+    [:fieldset.form-group
+     [:input.form-control.form-control-lg
+      {:placeholder placeholder
+       :on-change #(reset! form-data (assoc @form-data input-key (.. % -target -value)))
+       :on-blur validate-input!
+       :on-key-down validate-input!
+       :on-key-up validate-input!
+       :type (or input-type :text)
+       :value (get-in (rum/react form-data) [input-key])}]
+     (InputErrors (rum/react form-errors) input-key)]))
+
+(rum/defc UserFormBase <
+  {:will-unmount
+   (fn [{[r] :rum/args :as state}]
+     (reset! form-state initial-form-state)
```

Comment:
Shouldn't `form-state` be a local state of the component instead of a global one?